### DISCUSSION TITLE
[2.0] How to cleanup raw async connection?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#8979] [Usage Questions]
I had something like this previously working on 1.4 just fine with:
```
sample_engine = sa.create_engine("sqlite:///:memory:")
async def foo():
        engine = create_async_engine("sqlite:///:memory:", future=True)
        conn = await engine.raw_connection()
        sample_engine.raw_connection().backup(conn.connection)
        conn.close()
```

Now, on 2, I'm trying this:
```
sample_engine = sa.create_engine("sqlite:///:memory:")
async def foo():
        engine = create_async_engine("sqlite+aiosqlite:///:memory:")
        conn = await engine.raw_connection()
        sample_engine.raw_connection().backup(conn.driver_connection._conn)
        conn.close()
```

But, I get a bunch of errors like:
`sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?`
`RuntimeWarning: coroutine 'Connection.close' was never awaited`
`RuntimeWarning: coroutine 'Connection.rollback' was never awaited`

I also tried adding an `await`, but didn't help.

**Comment by zzzeek** [comment]
the code above produces the same set of errors on 1.4 as 2.0 for me.    The aiosqlite driver only works as an asyncio DBAPI so has to be called under a greenlet spawn.  this is usually a run_sync call but the "raw" call looks like

```py
import sqlalchemy as sa
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.util.concurrency import greenlet_spawn
import asyncio

sample_engine = sa.create_engine("sqlite:///:memory:")
async def foo():
    engine = create_async_engine("sqlite+aiosqlite:///:memory:")
    conn = await engine.raw_connection()
    sample_engine.raw_connection().backup(conn.driver_connection._conn)
    await greenlet_spawn(conn.close)


asyncio.run(foo())
```

I think the issue here is, if you want to run "backup", you'd just use the regular pysqlite connection.    you get no async benefits with the aiosqlite driver in any case.

**Reply by Dreamsorcerer** [reply]
I need an async engine returned for the application.

For context, this is used in the test setup for an aiohttp application. The intention is to create a sample DB for testing and then copy it fresh for each test. So, it doesn't matter that this is a blocking call here, but it does need an async engine at the end.

**Reply by zzzeek** [reply]
welp, greenlet_spawn is the secret sauce, you can use it for anything that raises that error, even ORM lazy loads.

**Reply by Dreamsorcerer** [reply]
OK, just figured out a proper solution. Seems like an issue going directly to `engine.raw_connection()`. Instead, I can go through a regular connection like this:
```
        async with engine.connect() as conn:
            raw = await conn.get_raw_connection()
            sample_engine.raw_connection().backup(raw.driver_connection._conn)
```

Everything works without warnings.

**Reply by zzzeek** [reply]
yes.  I think it's a little weird asyncengine has a raw_connection() method.
